### DISCUSSION TITLE
Can we use a C++ library from within a Zenoh plugin?

### DISCUSSION THREAD

**Discussion by PLeVasseur** [eclipse-zenoh/roadmap#124] [Q&A] [answered]
Hi there :wave: 

Was having an online discussion with @Mallets earlier today and wanted to make sure I understood correctly.

Related to some discussions occurring over on how to handle the architecture of our "transport bridge" component (called streamer) that bridges between various transports, e.g. Zenoh, SOME/IP for [uProtocol](https://github.com/eclipse-uprotocol).

We were discussing that using C++ code within a Zenoh plugin was not doable, but then it seemed to become it was possible when we clarified a bit further.

So I tried to write a little more detail around what we're trying to do. Is the following kind of architecture doable? 

```mermaid
flowchart TB
    subgraph ide0 [zenohd]
    main-->ide1
    subgraph ide1 [streamer-plugin]
    a1[up-streamer-rust]-->ide2
    a1[up-streamer-rust]-->ide3
    subgraph ide2 [OS thread]
    a2[up-client-zenoh-rust]
    end
    subgraph ide3 [OS thread]
    subgraph ide4 [Rust bridge e.g. up-client-someip-rust using cxx]
    a3[up-client-someip-cpp]
    end
    end
    end
    end
```

**Comment by evshary** [comment]
Hi @PLeVasseur 
I think what Luca means is we can use C++ to run in Zenoh router mode, and then load the plugins, which are written in Rust.
Maybe I can provide the example, hopefully tomorrow.
@Mallets Please correct me if I'm wrong.

**Reply by Mallets** [reply]
That's correct. A plugin has a Rust interface. Plugin code then can link other libraries (e.g. C/C++) but those libraries do not have access to the Zenoh C/C++ API since they are running in plugin-space.

**Reply by PLeVasseur** [reply]
Sure, I'm trying to piece it together :slightly_smiling_face: 

In this case though, the only C++ code is a C++ library wrapped in Rust code related to a different transport protocol, nothing to do with Zenoh, right?

Would the above architecture work in that case?

Or maybe I missed the point haha

**Reply by PLeVasseur** [reply]
@evshary -- I'm all for learning more about the approach you're outlining if @Mallets says what I'm proposing won't work

**Reply by Mallets** [reply, marked as answer]
Correct. An example of such approach is the [ROS2 plugin](https://github.com/eclipse-zenoh/zenoh-plugin-ros2dds) where the C-based CycloneDDS library is used to communicate over DDS. CycloneDDS though does not call anything related to Zenoh.

**Reply by PLeVasseur** [reply]
Okay, the picture is coming together.

So to be super clear and pedantic about this, [over here](https://github.com/eclipse-zenoh/zenoh-plugin-ros2dds/blob/b835413ebdc7ca4e4bdfb11b964bf5b6cf5fac36/zenoh-plugin-ros2dds/src/lib.rs#L242) in the ROS2 plugin we call `dds_create_participant` (wrapped in `unsafe` of course), which is the [same C function](https://github.com/eclipse-cyclonedds/cyclonedds/blob/79680a162494355a6991f474c956320a076b321a/src/core/ddsc/include/dds/dds.h#L1010) from the eclipse-cyclonedds project, right?

**Reply by JEnoch** [reply]
Correct: the CycloneDDS C code is wrapped into Rust in the [cyclors crate](https://github.com/ZettaScaleLabs/cyclors) using the [bindgen tool](https://github.com/rust-lang/rust-bindgen) in the [`build.rs`](https://github.com/ZettaScaleLabs/cyclors/blob/master/build.rs) file.
And the ROS2 plugin depends on cyclors.

**Reply by PLeVasseur** [reply]
Perfect! Thank you everyone for pitching in to help me understand :slightly_smiling_face: 

Only question is how the heck to Mark as answer multiple of them haha
